eventsources/sources/bitbucketserver/start.go

```diff
@@ -0,0 +1,354 @@
+
+	apiResponse, err := bitbucketClient.DefaultApi.FindWebhooks(bitbucketserverEventSource.ProjectKey, bitbucketserverEventSource.RepositorySlug, nil)
+
+	if err != nil {
+		return errors.Wrapf(err, "failed to list existing hooks to check for duplicates for repository %s/%s", router.bitbucketserverEventSource.ProjectKey, router.bitbucketserverEventSource.RepositorySlug)
+	}
+
+	hooks, err := bitbucketv1.GetWebhooksResponse(apiResponse)
+
+	if err != nil {
+		return errors.Wrapf(err, "failed to convert the list of webhooks for repository %s/%s", router.bitbucketserverEventSource.ProjectKey, router.bitbucketserverEventSource.RepositorySlug)
+	}
+
+	var existingHook bitbucketv1.Webhook
+	isAlreadyExists := false
+
+	for _, hook := range hooks {
+		if hook.Url == formattedURL {
+			isAlreadyExists = true
+			existingHook = hook
+			router.hookID = hook.ID
+			break
+		}
+	}
+
+	logger.Info("retrieving the webhook secret...")
+	webhookSecret, err := router.getWebhookSecret(bitbucketserverEventSource.WebhookSecret)
+	if err != nil {
+		return errors.Errorf("failed to get bitbucketserver webhook secret. err: %+v", err)
+	}
+
+	newHook := bitbucketv1.Webhook{
+		Name:          "Argo Events",
+		Url:           formattedURL,
+		Active:        true,
+		Events:        bitbucketserverEventSource.Events,
+		Configuration: bitbucketv1.WebhookConfiguration{Secret: webhookSecret.secret},
+	}
+
+	localVarPostBody, err := json.Marshal(newHook)
+	if err != nil {
+		return errors.Wrapf(err, "failed to marshal new webhook to JSON")
+	}
+
+	var createdHook *bitbucketv1.Webhook
+
+	// Create the webhook when it doesn't exist yet
+	if !isAlreadyExists {
+		apiResponse, err = bitbucketClient.DefaultApi.CreateWebhook(bitbucketserverEventSource.ProjectKey, bitbucketserverEventSource.RepositorySlug, localVarPostBody, []string{"application/json"})
+
+		if err != nil {
+			return errors.Errorf("failed to add webhook. err: %+v", err)
+		}
+
+		err = mapstructure.Decode(apiResponse.Values, &createdHook)
+		if err != nil {
+			return errors.Errorf("failed to convert API response to Webhook struct. err: %+v", err)
+		}
+		router.hookID = createdHook.ID
+
+		logger.With("hook-id", createdHook.ID).Info("hook succesfully registered")
+	}
+
+	// Update the webhook when it does exist and the configuration has chagned
+	if isAlreadyExists && (!reflect.DeepEqual(existingHook.Events, newHook.Events) || !reflect.DeepEqual(existingHook.Configuration, newHook.Configuration)) {
+		logger.Info("webhook already exists and configuration has changed. Updating webhook.")
+
+		apiResponse, err = bitbucketClient.DefaultApi.UpdateWebhook(bitbucketserverEventSource.ProjectKey, bitbucketserverEventSource.RepositorySlug, int32(existingHook.ID), localVarPostBody, []string{"application/json"})
+
+		if err != nil {
+			return errors.Errorf("failed to update webhook. err: %+v", err)
+		}
+
+		err = mapstructure.Decode(apiResponse.Values, &createdHook)
```

Comment:
It looks like this part can be skipped, as it's an update, and the hook id is already there.

Reply:
You're right. Fixed it. Thanks!